### PR TITLE
Feature/tcp socket select

### PR DESCRIPTION
**Pull request by GuyLewin**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

Implements #1620

### CODE REVIEW

**Comment by GuyLewin** [comment]
@spacewander do you mind looking at this too? This PR is a lot more complex than the other, but I believe it's a *very* important feature for us to have (for example in proxying web sockets or other TCP sockets).

**Comment by spacewander** [comment]
Does this new method have any advantage over running multiple coroutines (created by `ngx.thread.spawn`, each one contains their own sockets) and sync with `ngx.semaphore`? The latter one is already battle-tested.

File: src/ngx_http_lua_socket_tcp.c [hunk 1]
```diff
@@ -2927,6 +3086,83 @@ ngx_http_lua_socket_tcp_send_retval_handler(ngx_http_request_t *r,
...
+        }
+
+        lua_pushnumber(L, matched_index + 1);
+        (void) ngx_http_lua_socket_read_error_retval_handler(r, u, L);
+        lua_remove(L, -2);
+        return 2;
+    }
+
+    lua_pushnumber(L, matched_index + 1);
+    return 1;
```
Comment (spacewander) [inline src/ngx_http_lua_socket_tcp.c]: I am wonder if returning only one readable socket is OK... Please correct me if I am wrong:
At least with epoll event module, the read event is triggered only once at the edge (search EPOLLET as the keyword). If multiple events happen at the same time, and only one event is handled, the other shots will be lost.

**Comment by GuyLewin** [comment]
@spacewander thanks for going over this! Can you give an example for Lua code you'd suggest to replace this functionality? Obviously I'd want this code to run without extra threads, but all based on the epoll() (or any module)'s multiplexing logic.

File: src/ngx_http_lua_socket_tcp.c (hunk 1 above)
Comment (GuyLewin) [reply src/ngx_http_lua_socket_tcp.c]: You're definitely right. I'm not sure how conceptually I could do that (combine all the events from the same epoll into 1 return code) but I'll look into that.
I believe that even if a solution of returning multiple FDs doesn't work out - this select() API is very relevant, as you can call select() again after you just called it and get the remaining FD(s).
I'll look into it and update.

**Comment by spacewander** [comment]
@GuyLewin 
You can use the example in `ngx.semaphore` docs as a start:
https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/semaphore.md

Personally speaking, I think the coroutines could be cheap enough. And implementing a new layer of I/O multiplexing (in the Lua code) above the other one (in the Nginx) may be too complex and inflexible. Somebody has used epoll to reduce memory usage in Go: https://www.freecodecamp.org/news/million-websockets-and-go-cc58418460bb, but they called the epoll directly so this is a different story.

**Comment by zhuizhuhaomeng** [comment]
I don't want to merge this feature. I believe the existing features are already sufficient to meet the requirements.

**Pull request closed without merging**